File: 算法/排序算法/QuickSort.java
/**
 * 快速排序算法-从小到大
 * yangshengkai
 * 时间复杂度-最佳：平均：最差 = O(n logn):O(n logn):O(n^2)
 * 空间复杂度-O(logn)
 * 不稳定排序
 */
class QuickSort{

    //排序算法
    public static void quickSoftAlg(int[] arr, int left, int right){
        int i = left, j = right, temp;
        int middle;

        middle = arr[(i + j) / 2];
        do{
            //从左查找比中间值大的数
            while (arr[i] < middle && i < right){
                i++;
            }
            //从右查找比中间值小数
            while (arr[j] > middle && j > left){
                j--;
            }
            //交换两边的数值
            if (i <= j){
                temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
                i++;
                j--;
            }
        }while (i <= j);

        if (i < right){
            quickSoftAlg(arr, i, right);
        }
        if (j > left){
            quickSoftAlg(arr, left, j);
        }

    }
    //主方法
    public static void main(String[] str){
        QuickSort quickSort = new QuickSort();
        int[] testArr = new int[]{7,9,6,1,3,65,68,2,18,96,23,15};
        quickSort.quickSoftAlg(testArr, 0, testArr.length - 1);
        for (int m = 0; m < testArr.length; m++){
            System.out.print(testArr[m] + " ");
        }
    }
}

File: Algorithms/PHP/sort/QuickSort.php
<?php
/**
 * 快速排序-不稳定
 * 时间复杂度-O(nlogn)，空间复杂度-O(1)
 * @author skyeinfo@qq.com
 * @lastModifyTime 2018/11/23
 * @lastModify skyeinfo@qq.com
 */
class QuickSort
{
    public function mainOne(array &$sortArr, $left, $right) {

        $i = $left;
        $j = $right;
        $middle = $sortArr[($i + $j) / 2];

        do {
            while ($sortArr[$i] < $middle && $i < $right) {
                $i++;
            }

            while ($sortArr[$j] > $middle && $j > $left) {
                $j--;
            }
            //var_dump($i,$j);
            if ($i <= $j) {
                $temp = $sortArr[$i];
                $sortArr[$i] = $sortArr[$j];
                $sortArr[$j] = $temp;

                $i++;
                $j--;
            }
        } while($i <= $j);

        if ($i < $right) {
            $this->mainOne($sortArr, $i, $right);
        }

        if ($j > $left) {
            $this->mainOne($sortArr, $left, $j);
        }
    }

    /**
     * 分解的过程可解Top-K问题
     * @author skyeinfo@qq.com
     * @lastModifyTime 2018/11/23
     * @lastModify skyeinfo@qq.com
     * @param array $sortArr
     * @param $start
     * @param $end
     */
    public function mainTwo(array &$sortArr, $start, $end) {
        if ($start >= $end) return;

        $pivot = $sortArr[$end];
        $i = $start;

        for ($j = $start; $j < $end; $j++) {
            if ($sortArr[$j] < $pivot) {
                $temp = $sortArr[$i];
                $sortArr[$i] = $sortArr[$j];
                $sortArr[$j] = $temp;

                $i++;
            }
        }

        $temp = $sortArr[$end];
        $sortArr[$end] = $sortArr[$i];
        $sortArr[$i] = $temp;

        $this->mainTwo($sortArr, $start, $i - 1);
        $this->mainTwo($sortArr, $i + 1, $end);
    }
}

$a = new QuickSort();

$arr = [15,12,4,24,5,21,6,32,56,14,37,23];
$a->mainOne($arr, 0, count($arr) - 1);
print_r($arr);

$arr = [15,12,4,24,5,21,6,32,56,14,37,23];
$a->mainTwo($arr, 0, count($arr) - 1);
print_r($arr);